src/Init/Lean/PrettyPrinter/Parenthesizer.lean

```diff
@@ -0,0 +1,444 @@
+let c := Expr.constName? p.getAppFn;
+env ← liftM getEnv;
+match c >>= (parenthesizerAttribute.ext.getState env).table.find? with
+| some (f::_) => do
+  -- call first matching parenthesizer
+  f p
+| _           => do
+  -- (try to) unfold definition and recurse
+  some p' ← liftM $ unfoldDefinition? p
+    | throw $ Exception.other $ "no known parenthesizer for '" ++ toString p ++ "'";
+  visit p'
+
+open Lean.Parser
+
+-- Macro scopes in the parenthesizer output are ultimately ignored by the pretty printer,
+-- so give a trivial implementation.
+instance monadQuotation : MonadQuotation ParenthesizerM := {
+  getCurrMacroScope   := pure $ arbitrary _,
+  getMainModule       := pure $ arbitrary _,
+  withFreshMacroScope := fun α x => x,
+}
+
+/-- Recurse using `visit`, and parenthesize the result using `mkParen` if necessary. -/
+def visitParenthesizable (mkParen : Syntax → Syntax) (rbp : Nat) : ParenthesizerM Unit := do
+stx ← getCur;
+idx ← getIdx;
+⟨t, lbp, trailRbp⟩ ← get;
+-- reset lbp/rbp and store `mkParen` for the recursive call
+set { stxTrav := t };
+adaptReader (fun (ctx : Context) => { ctx with mkParen := some mkParen }) $
+  -- we assume that each node kind is produced by a 0-ary parser of the same name
+  visit (mkConst stx.getKind);
+⟨_, some lbp', trailRbp'⟩ ← get
+  | panic! "visitParenthesizable: visited a term without tokens?!";
+trace! `PrettyPrinter.parenthesize ("...precedences are " ++ fmt rbp ++ " >? " ++ fmt lbp' ++ ", " ++ fmt trailRbp' ++ " <? " ++ fmt lbp);
+-- Should we parenthesize?
+trailRbp' ← if rbp > lbp' || (match trailRbp', lbp with some trailRbp', some lbp => trailRbp' < lbp | _, _ => false) then do
+    -- The recursive `visit` call, by the invariant, has moved to the next node to the left. In order to parenthesize
+    -- the original node, we must first move to the right, except if we already were at the left-most child in the first
+    -- place.
+    when (idx > 0) goRight;
+    setCur (mkParen stx);
+    goLeft;
+    -- after parenthesization, there is no more trailing parser
+    pure none
+  else pure trailRbp';
+-- If we already had a token at this level (`lbp ≠ none`), keep the trailing parser. Otherwise, use the minimum of
+-- `rbp` and `trailRbp'`.
+let trailRbp := match trailRbp', lbp with
+  | _,              some _ => trailRbp
+  | some trailRbp', _      => some (Nat.min trailRbp' rbp)
+  | _,              _      => some rbp;
+modify (fun st => { st with lbp := lbp' <|> lbp, trailRbp := trailRbp })
+
+/-- Set token precedence and advance to the left. -/
+def visitToken (lbp : Nat) : ParenthesizerM Unit := do
+modify (fun st => { st with lbp := lbp });
+goLeft
+
+/-- Execute `x` at the right-most child of the current node, if any, then advance to the left. -/
+def visitArgs (x : ParenthesizerM Unit) : ParenthesizerM Unit := do
+stx ← getCur;
+when (stx.getArgs.size > 0) $
+  goDown (stx.getArgs.size - 1) *> x <* goUp;
```

Comment:
I am wondering whether it is worth adding `curHasChild` and `goRightMostChild` methods. Then, we can write
```lean
whenM curHasChild $ 
  goRightMostChild *> x <* goUp;
```
Not sure. Just a random comment.

Reply:
Let's see if we need them more often in the future